src/TriggerBinding/SqlTableChangeMonitor.cs

```diff
@@ -0,0 +1,774 @@
+                            {
+                                await updateTablesPreInvocationCommand.ExecuteNonQueryAsync(token);
+                            }
+
+                            // Use the version number to query for new changes.
+                            using (SqlCommand getChangesCommand = this.BuildGetChangesCommand(connection, transaction))
+                            {
+                                var rows = new List<IReadOnlyDictionary<string, string>>();
+                                using (SqlDataReader reader = await getChangesCommand.ExecuteReaderAsync(token))
+                                {
+                                    while (await reader.ReadAsync(token))
+                                    {
+                                        rows.Add(SqlBindingUtilities.BuildDictionaryFromSqlRow(reader));
+                                    }
+                                }
+
+                                this._rows = rows;
+                            }
+
+                            // If changes were found, acquire leases on them.
+                            if (this._rows.Count > 0)
+                            {
+                                using (SqlCommand acquireLeasesCommand = this.BuildAcquireLeasesCommand(connection, transaction))
+                                {
+                                    await acquireLeasesCommand.ExecuteNonQueryAsync(token);
+                                }
+                            }
+                            transaction.Commit();
+                        }
+                        catch (Exception ex)
+                        {
+                            this._logger.LogError("Commit Exception Type: {0}", ex.GetType());
+                            this._logger.LogError("  Message: {0}", ex.Message);
+                            try
+                            {
+                                transaction.Rollback();
+                            }
+                            catch (Exception ex2)
+                            {
+                                this._logger.LogError("Rollback Exception Type: {0}", ex2.GetType());
+                                this._logger.LogError("  Message: {0}", ex2.Message);
+                            }
+                        }
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                // If there's an exception in any part of the process, we want to clear all of our data in memory and
+                // retry checking for changes again.
+                this._rows = new List<IReadOnlyDictionary<string, string>>();
+                this._logger.LogWarning($"Failed to check {this._userTable.FullName} for new changes due to error: {e.Message}");
+            }
+        }
+
+        private async Task ProcessChangesAsync(CancellationToken token)
```

Comment:
I may be missing something here, but won't the item get re-picked after the current lease expires? The worker will need to definitely acquire lease before processing and there's no easy way to revert the changes made to the worker-table if the processing fails so if the item gets available again after lease period expires, perhaps that is an acceptable behavior to live with?

Reply:
I think the main problem is that we update the change version when we take out the lease instead of only after it's been processed.

This is something we should follow up on - please make an issue to track this and we can take a deeper look at it later.